Sources/core/objects/Projectiles.cpp

```diff
@@ -0,0 +1,321 @@
+#include "Projectiles.h"
+
+#include "../Helpers.h"
+#include "Mob.h"
+
+#include "../Game.h"
+
+#include <assert.h>
+
+Projectile::Projectile(size_t id) : IOnMapObject(id)
+{
+    lastMove = 0;
+    SetPassable(D_ALL, Passable::AIR);
+    tickSpeed = 0.25;
+    pixSpeed = 1;
+    dMove = D_DOWN;
+    anchored = false;
+    damage = 0;
+
+    v_level = 5;
+    SetSprite("icons/projectiles.dmi");
+}
+
+bool Projectile::TryMove(Dir direct)
+{
+    if (!CheckMoveTime())
+    {
+        return false;
+    }
+    if (anchored)
+    {
+        return false;
+    }
+    if (!Rotate(direct))
+    {
+        return false;
+    }
+    if (!CheckPassable())
+    {
+        return false;
+    }
+    return MainMove();
+}
+
+void Projectile::ProcessMovement()
+{
+    int z = current_step_;
+    Dir step = VDirToDir(movement_[z]);
+
+    TryMove(step);
+    current_step_++;
+    if(!(current_step_<movement_.size()) && !reached_target )
+    {
+        current_step_ = 0;
+        VDir temp = movement_[0];
+        movement_.clear();
+        reached_target = true;
+        movement_.push_back(temp);
+    }
+    if(!(current_step_<movement_.size()) && reached_target )
+    {
+        current_step_ = 0;
+    }    
+}
+
+bool Projectile::CheckMoveTime()
+{
+    if ((static_cast<int>(MAIN_TICK) - lastMove) < tickSpeed)
+    {
+        return false;
+    }
+    return true;
+}
+
+bool Projectile::CheckPassable()
+{
+    PassableLevel loc = GetPassable(dMove);
+    if (loc != Passable::FULL)
+    {
+        SetPassable(dMove, Passable::FULL);
+    }
+    if (!CanPass(owner->GetPassable(dMove), passable_level))
+    {
+        owner->Bump(GetId());
+        if (loc != Passable::FULL)
+        {
+            SetPassable(dMove, loc);
+        }
+        return false;
+    }
+    if (loc != Passable::FULL)
+    {
+        SetPassable(dMove, loc);
+    }
+
+    auto neighbour = owner->GetNeighbour(dMove);
+    if (   !CanPass(neighbour->GetPassable(D_ALL), passable_level)
+        || !CanPass(neighbour->GetPassable(helpers::revert_dir(dMove)), passable_level))
+    {
+        neighbour->Bump(GetId());
+        return false;
+    }
+    
+    return true;
+}
+
+bool Projectile::Rotate(Dir dir)
+{
+    dMove = dir;
+    return true;
+}
+
+bool Projectile::MainMove()
+{
+    auto new_owner = owner->GetNeighbour(dMove);
+    if (new_owner == owner)
+    {
+        return false;
+    }
+
+    owner->RemoveItem(GetId());
+    new_owner->AddItem(GetId());
+
+    lastMove = static_cast<int>(MAIN_TICK);
+    return true;
+}
+
+
+void Projectile::Bump(id_ptr_on<IMovable> item)
+{
```

Comment:
That code is the main problem. 
For some reason you delete global game class - you should not do this.
If you would like to remove some object you should just call `Delete()` function of this object.
Also, the function `Bump()` is called from a subject of bumping (e.g, when human run into wall the function `Bump` of the wall will be called, not the human `Bump`).